File: Practicals/prac_4_3.py
import os

def main():
    pid = os.fork()  # Fork to create a new process

    if pid == 0:
        # Child process: executing the same program but with different code
        print(f"Child process, PID: {os.getpid()} executing 'ls -l'")
        # Replace the child process with 'ls -l' command
        os.execvp("ls", ["ls", "-l"])
    else:
        # Parent process
        print(f"Parent process, PID: {os.getpid()} - Waiting for child to finish")
        os.wait()  # Parent waits for the child to finish
        print("Parent Task Done, Child has finished")

if __name__ == "__main__":
    main()

File: Practicals/prac_4_3.cpp
#include <iostream>
#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>

int main() {
    pid_t pid = fork(); // Create a new process

    if (pid < 0) {
        std::cerr << "Fork failed\n";
        return 1;
    }

    if (pid == 0) { 
        // Child process
        std::cout << "Child Process, PID: " << getpid() << "\n";
        sleep(2); // Simulate work
        std::cout << "Child Task Done\n";
    } 
    else { 
        // Parent process
        std::cout << "Parent Process, PID: " << getpid() << "\n";
        wait(NULL); // Wait for the child to complete
        std::cout << "Parent Task Done\n";
    }

    return 0;
}
